File: pirates-ios/PiratesGame/Components/Game/LevelView.swift
//
//  LevelView.swift
//  Game view: a level
//

import UIKit
import UniLayout
import JsonInflator

class LevelView: FrameContainerView {

    // --
    // MARK: Members
    // --
    
    private let tileMapView = LevelTileMapView()
    private let entitiesView = LevelEntitiesView()
    private let waveAnimationView = LevelWaveAnimationView()


    // --
    // MARK: Viewlet integration
    // --
    
    override class func viewlet() -> JsonInflatable {
        return ViewletClass()
    }
    
    private class ViewletClass: JsonInflatable {
        
        func create() -> Any {
            return LevelView()
        }
        
        func update(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any], parent: Any?, binder: InflatorBinder?) -> Bool {
            if let levelView = object as? LevelView {
                // Apply animation
                levelView.waveSpawnInterval = convUtil.asFloat(value: attributes["waveSpawnInterval"]) ?? 0.2
                
                // Set tiles
                levelView.tileMap = convUtil.asStringArray(value: attributes["tileMap"])

                // Generic view properties
                ViewletUtil.applyGenericViewAttributes(convUtil: convUtil, view: levelView, attributes: attributes)

                // Event handling
                levelView.tapEvent = AppEvent(value: attributes["tapEvent"])

                // Forward event observer
                if let eventObserver = parent as? AppEventObserver {
                    levelView.eventObserver = eventObserver
                }
                return true
            }
            return false
        }
        
        func canRecycle(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any]) -> Bool {
            return object is LevelView
        }
        
    }


    // --
    // MARK: Initialization
    // --

    override init(frame: CGRect) {
        super.init(frame: frame)
        setup()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        setup()
    }
    
    fileprivate func setup() {
        // Add tile map view
        tileMapView.layoutProperties.width = UniLayoutProperties.stretchToParent
        tileMapView.layoutProperties.height = UniLayoutProperties.stretchToParent
        addSubview(tileMapView)
        
        // Add wave animation view
        waveAnimationView.layoutProperties.width = UniLayoutProperties.stretchToParent
        waveAnimationView.layoutProperties.height = UniLayoutProperties.stretchToParent
        waveAnimationView.tileMapView = tileMapView
        addSubview(waveAnimationView)
        
        // Add entities view
        entitiesView.layoutProperties.width = UniLayoutProperties.stretchToParent
        entitiesView.layoutProperties.height = UniLayoutProperties.stretchToParent
        addSubview(entitiesView)
        
        // Start spawning wave particles
        continueWaveSpawning()
    }
    

    // --
    // MARK: Configurable values
    // --
    
    var waveSpawnInterval: Float = 0.2
    
    var tileMap: [String] {
        set {
            tileMapView.tiles = newValue
            entitiesView.tiles = newValue
        }
        get { return tileMapView.tiles }
    }
   

    // --
    // MARK: Spawning wave particles
    // --
    
    private func continueWaveSpawning() {
        DispatchQueue.main.asyncAfter(deadline: .now() + Double(waveSpawnInterval), execute: { [weak self] in
            if let level = self {
                level.waveAnimationView.spawnRandomWave()
                level.continueWaveSpawning()
            }
        })
    }

}

File: pirates-android/app/src/main/java/com/crescentflare/piratesgame/components/game/LevelView.kt
package com.crescentflare.piratesgame.components.game

import android.annotation.TargetApi
import android.content.Context
import android.os.Build
import android.util.AttributeSet
import android.view.ViewGroup
import com.crescentflare.piratesgame.components.containers.FrameContainerView
import com.crescentflare.piratesgame.components.game.levelhelpers.LevelEntitiesView
import com.crescentflare.piratesgame.components.game.levelhelpers.LevelTileMapView
import com.crescentflare.piratesgame.components.game.levelhelpers.LevelWaveAnimationView
import com.crescentflare.piratesgame.components.utility.ViewletUtil
import com.crescentflare.piratesgame.infrastructure.events.AppEvent
import com.crescentflare.piratesgame.infrastructure.events.AppEventObserver
import com.crescentflare.unilayout.helpers.UniLayoutParams
import com.crescentflare.jsoninflator.JsonInflatable
import com.crescentflare.jsoninflator.binder.InflatorBinder
import com.crescentflare.jsoninflator.utility.InflatorMapUtil
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/**
 * Game view: a level
 */
class LevelView : FrameContainerView {

    // --
    // Static: viewlet integration
    // --

    companion object {

        val viewlet: JsonInflatable = object : JsonInflatable {

            override fun create(context: Context): Any {
                return LevelView(context)
            }

            override fun update(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>, parent: Any?, binder: InflatorBinder?): Boolean {
                if (obj is LevelView) {
                    // Apply animation
                    obj.waveSpawnInterval = mapUtil.optionalFloat(attributes, "waveSpawnInterval", 0.2f)

                    // Set tiles
                    obj.tileMap = mapUtil.optionalStringList(attributes, "tileMap")

                    // Generic view properties
                    ViewletUtil.applyGenericViewAttributes(mapUtil, obj, attributes)

                    // Event handling
                    obj.tapEvent = AppEvent.fromObject(attributes["tapEvent"])

                    // Forward event observer
                    if (parent is AppEventObserver) {
                        obj.eventObserver = parent
                    }
                    return true
                }
                return false
            }

            override fun canRecycle(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>): Boolean {
                return obj is LevelView
            }

        }

    }


    // --
    // Members
    // --

    private val tileMapView: LevelTileMapView
    private val entitiesView: LevelEntitiesView
    private val waveAnimationView: LevelWaveAnimationView
    private var isAttached = false
    private var spawnIdle = true


    // --
    // Initialization
    // --

    @JvmOverloads
    constructor(
            context: Context,
            attrs: AttributeSet? = null,
            defStyleAttr: Int = 0)
            : super(context, attrs, defStyleAttr)

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    constructor(
            context: Context,
            attrs: AttributeSet?,
            defStyleAttr: Int,
            defStyleRes: Int)
            : super(context, attrs, defStyleAttr, defStyleRes)

    init {
        // Add tile map view
        tileMapView = LevelTileMapView(context)
        tileMapView.layoutParams = UniLayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
        addView(tileMapView)

        // Add wave animation view
        waveAnimationView = LevelWaveAnimationView(context)
        waveAnimationView.layoutParams = UniLayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
        waveAnimationView.tileMapView = tileMapView
        addView(waveAnimationView)

        // Add entities view
        entitiesView = LevelEntitiesView(context)
        entitiesView.layoutParams = UniLayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
        addView(entitiesView)

        // Start spawning wave particles
        continueWaveSpawning()
    }


    // --
    // Configurable values
    // --

    var waveSpawnInterval = 0.2f

    var tileMap: MutableList<String>
        set(tileMap) {
            tileMapView.tiles = tileMap
            entitiesView.tiles = tileMap
        }
        get() = tileMapView.tiles


    // --
    // Spawning wave particles
    // --

    private fun continueWaveSpawning() {
        spawnIdle = false
        GlobalScope.launch(Dispatchers.Main) {
            delay((waveSpawnInterval * 1000).toLong())
            waveAnimationView.spawnRandomWave()
            spawnIdle = true
            if (isAttached) {
                continueWaveSpawning()
            }
        }
    }


    // --
    // Listen for attachment/detachment
    // --

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        isAttached = false
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        isAttached = true
        if (spawnIdle) {
            continueWaveSpawning()
        }
    }

}
